frontend/projects/upgrade/src/app/core/stratification-factors/store/stratification-factors.effects.ts

```diff
@@ -0,0 +1,109 @@
+import { Injectable } from '@angular/core';
+import { Router } from '@angular/router';
+import { Actions, createEffect, ofType } from '@ngrx/effects';
+import { map, switchMap, filter, catchError, withLatestFrom } from 'rxjs/operators';
+import { Store, select } from '@ngrx/store';
+import { AppState } from '../../core.state';
+import * as StratificationFactorsActions from './stratification-factors.actions';
+import { StratificationFactorsDataService } from '../stratification-factors.data.service';
+import { selectAllStratificationFactors } from './stratification-factors.selectors';
+import { StratificationFactor } from './stratification-factors.model';
+
+@Injectable()
+export class StratificationFactorsEffects {
+  constructor(
+    private actions$: Actions,
+    private router: Router,
+    private store$: Store<AppState>,
+    private stratificationFactorsDataService: StratificationFactorsDataService
+  ) {}
+
+  fetchStratificationFactors$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(StratificationFactorsActions.actionFetchStratificationFactors),
+      withLatestFrom(this.store$.pipe(select(selectAllStratificationFactors))),
+      switchMap(() =>
+        this.stratificationFactorsDataService.fetchStratificationFactors().pipe(
+          map((data: StratificationFactor[]) =>
+            StratificationFactorsActions.actionFetchStratificationFactorsSuccess({
+              stratificationFactors: data,
+            })
+          ),
+          catchError(() => [StratificationFactorsActions.actionFetchStratificationFactorsFailure()])
+        )
+      )
+    )
+  );
+
+  deleteStratificationFactor$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(StratificationFactorsActions.actionDeleteStratificationFactor),
+      map((action) => action.factor),
+      filter((factor) => !!factor),
+      switchMap((factor) =>
+        this.stratificationFactorsDataService.deleteStratificationFactor(factor).pipe(
+          map((data) => {
+            if (data[0]) {
```

Comment:
If `data` comes back successfully but `data[0]` is unexpectedly falsey, what should happen? I think this would return whatever this unexpected `data` data is and not navigate anywhere, which is probably not wanted.

I guess what should happen would be same as the `catchError` scenarios, `StratificationFactorsActions.actionDeleteStratificationFactorFailure()`, but does failure also need to do `router.navigate`?

Reply:
Giving data a type there will resolve the first point you mentioned right? 
Regarding navigation don't you think success or failure in both cases we should route back to stratification tab and not at home page?

Reply:
what i'm saying is that this is an `if` block without an else. So in the case where this is "successful" response from the db call but `data[0]` is null or undefined (such as if the `data` is an empty array), there is no `else` block to handle what to do.

Reply:
I tried running the unit tests for `deleteStratificationFactor$` to see if this was covered, but they are not really working; try making them fail, they will keep passing. I'll put the reason why in the comments for tests.

Reply:
Okay, got it. Will resolve this.